tools/audio/sampleconv/src/container/wav.c

```diff
@@ -0,0 +1,533 @@
+    uint16_t num_channels;
+    uint32_t sample_rate;
+    uint32_t byte_rate;
+    uint16_t block_align;
+    uint16_t bit_depth;
+} wav_fmt;
+
+typedef struct {
+    uint32_t num_samples;
+} wav_fact;
+
+typedef struct {
+    int8_t base_note;
+    int8_t fine_tune;
+    int8_t gain;
+    int8_t key_low;
+    int8_t key_hi;
+    int8_t vel_low;
+    int8_t vel_hi;
+    char _pad[1];
+} wav_inst;
+
+typedef struct {
+    uint32_t manufacturer;
+    uint32_t product;
+    uint32_t sample_period;
+    uint32_t unity_note;
+    char _pad[3];
+    uint8_t fine_tune;
+    uint32_t format;
+    uint32_t offset;
+    uint32_t num_sample_loops;
+    uint32_t sampler_data;
+} wav_smpl;
+
+typedef struct {
+    uint32_t cue_point_index;
+    uint32_t type;
+    uint32_t start;
+    uint32_t end;
+    uint32_t fraction;
+    uint32_t num;
+} wav_loop;
+
+int
+wav_read(container_data *out, const char *path, UNUSED bool matching)
+{
+    bool has_fmt = false;
+    bool has_fact = false;
+    bool has_data = false;
+    bool has_inst = false;
+    bool has_smpl = false;
+    memset(out, 0, sizeof(*out));
+
+    FILE *in = fopen(path, "rb");
+    if (in == NULL)
+        error("Failed to open \"%s\" for reading", path);
+
+    char riff[4];
+    uint32_t size;
+    char wave[4];
+
+    FREAD(in, riff, 4);
+    FREAD(in, &size, 4);
+    size = le32toh(size);
+    FREAD(in, wave, 4);
+
+    if (!CC4_CHECK(riff, "RIFF") || !CC4_CHECK(wave, "WAVE"))
+        error("Not a wav file?");
+
+    while (true) {
+        bool skipped = false;
+
+        long start = ftell(in);
+        if (start > 8 + size) {
+            error("Overran file");
+        }
+        if (start == 8 + size) {
+            break;
+        }
+
+        char cc4[4];
+        uint32_t chunk_size;
+
+        FREAD(in, cc4, 4);
+        FREAD(in, &chunk_size, 4);
+        chunk_size = le32toh(chunk_size);
+
+        switch (CC4(cc4[0], cc4[1], cc4[2], cc4[3])) {
+            case CC4('f', 'm', 't', ' '): {
+                wav_fmt fmt;
+                FREAD(in, &fmt, sizeof(fmt));
+                fmt.type = le16toh(fmt.type);
+                fmt.num_channels = le16toh(fmt.num_channels);
+                fmt.sample_rate = le32toh(fmt.sample_rate);
+                fmt.byte_rate = le32toh(fmt.byte_rate);
+                fmt.block_align = le16toh(fmt.block_align);
+                fmt.bit_depth = le16toh(fmt.bit_depth);
+
+                if (fmt.bit_depth != 16)
+                    error("Wav input format should be 16-bit PCM, was %u-bit", fmt.bit_depth);
+
+                switch (fmt.type) {
+                    case WAVE_TYPE_PCM:
+                        out->data_type = SAMPLE_TYPE_PCM16;
+                        break;
+
+                    case WAVE_TYPE_FLOAT:
+                    case WAVE_TYPE_MULAW:
+                    case WAVE_TYPE_ALAW:
+                    case WAVE_TYPE_EXTENSIBLE:
+                        error("Unhandled sample type: %d, should be PCM", fmt.type);
```

Comment:
is this is going to be a common error, maybe the type could be reported as a googleable string instead of an integer?